syntax = "proto3";

service GameCore {
  rpc MakeSession(MakeSessionRequest) returns (SessionID);
  rpc JoinSession(JoinRequest) returns (PlayerInfo);
  rpc DoLeave(LeaveRequest) returns (CoreResponse);
  rpc DoChat(ChatRequest) returns (CoreResponse);
  rpc DoVoteSacrifice(SacrificeRequest) returns (CoreResponse);
  rpc DoVoteMurder(MurderRequest) returns (CoreResponse);
  rpc DoGetStatus(StatusRequest) returns (CoreResponse);
}

message MakeSessionRequest { optional string session_id = 1; }

message JoinRequest {
  string player_name = 1;
  string session_id = 2;
}

message LeaveRequest { PlayerInfo key = 1; }

message StatusRequest { PlayerInfo key = 1; }

message ChatRequest {
  PlayerInfo key = 1;
  string message = 2;
}

message MurderRequest {
  PlayerInfo key = 1;
  string target_name = 2;
}

message SacrificeRequest {
  PlayerInfo key = 1;
  string target_name = 2;
}

message SessionID { string session_id = 1; }

message CoreResponse { string message = 1; }

message PlayerInfo {
  string player_token = 1;
  string session_id = 2;
}
